packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol

```diff
@@ -61,27 +335,325 @@ library ConnextUtils {
+   * @param _recipient The address that will receive the liquidity being removed
+   */
+  function removeLiquidity(
+    uint256 _amount,
     address _local,
+    address _recipient,
+    mapping(address => mapping(address => uint256)) storage _routerBalances,
+    IWrapped _wrapper
+  ) external {
+    // Sanity check: to is sensible
+    if (_recipient == address(0)) revert ConnextUtils__removeLiquidity_recipientEmpty();
+
+    // Sanity check: nonzero amounts
+    if (_amount == 0) revert ConnextUtils__removeLiquidity_amountIsZero();
+
+    uint256 routerBalance = _routerBalances[msg.sender][_local];
+    // Sanity check: amount can be deducted for the router
+    if (routerBalance < _amount) revert ConnextUtils__removeLiquidity_insufficientFunds();
+
+    // Update router balances
+    unchecked {
+      _routerBalances[msg.sender][_local] = routerBalance - _amount;
+    }
+
+    // Transfer from contract to specified to
+    _transferAssetFromContract(_local, _recipient, _amount, _wrapper);
+
+    // Emit event
+    emit LiquidityRemoved(msg.sender, _recipient, _local, _amount, msg.sender);
+  }
+
+  function xcall(
+    xCallLibArgs calldata _args,
+    mapping(address => ConnextMessage.TokenId) storage _adoptedToCanonical,
+    mapping(bytes32 => IStableSwap) storage _adoptedToLocalPools,
+    mapping(bytes32 => uint256) storage _relayerFees
+  ) external returns (bytes32, uint256) {
+    _xcallSanityChecks(_args);
+
+    // get the true transacting asset id (using wrapped native instead native)
+    (bytes32 transferId, bytes memory message, xCalledEventArgs memory eventArgs) = _processXcall(
+      _args,
+      _adoptedToCanonical,
+      _adoptedToLocalPools
+    );
+
+    // Store the relayer fee
+    _relayerFees[transferId] = _args.xCallArgs.relayerFee;
+
+    // emit event
+    emit XCalled(transferId, _args.xCallArgs, eventArgs, _args.nonce, message, msg.sender);
+
+    return (transferId, _args.nonce + 1);
+  }
+
+  /**
+   * @notice Called on the destination domain to disburse correct assets to end recipient
+   * and execute any included calldata
+   * @dev Can be called prior to or after `handle`, depending if fast liquidity is being
+   * used.
+   */
+  function execute(
+    ExecuteLibArgs calldata _args,
+    mapping(bytes32 => address[]) storage _routedTransfers,
+    mapping(bytes32 => bool) storage _reconciledTransfers,
+    mapping(address => mapping(address => uint256)) storage _routerBalances,
+    mapping(bytes32 => IStableSwap) storage _adoptedToLocalPools,
+    mapping(bytes32 => address) storage canonicalToAdopted,
+    RouterPermissionsManagerInfo storage _routerInfo,
+    mapping(bytes32 => address) storage _transferRelayer
+  ) external returns (bytes32 transferId) {
+    (bytes32 transferId, bool reconciled) = _executeSanityChecks(_args, _transferRelayer, _reconciledTransfers);
+
+    // check to see if the transfer was reconciled
+    (uint256 amount, address adopted) = _handleExecuteLiquidity(
```

Comment:
Note: the only thing this method does if the transfer is `reconciled` (meaning if the transfer is not a fast-liq transfer) is literally just call `_swapFromLocalAssetIfNeeded`

We may want to optimize this by removing the `isFast` boolean input to this function and literally deciding whether or not to call it here, as the consumer

And then calling `_swapFromLocalAssetIfNeeded` separately, using the result.

As in:
```solidity
if (reconciled) {
  (uint256 amount, address adopted) = _swapFromLocalAssetIfNeeded(...);
} else {
  (uint256 amount, address adopted) = _handleExecuteLiquidity(
        transferId,
        // no `!reconciled` argument!
      _args,
      _routedTransfers,
      _routerBalances,
      _adoptedToLocalPools,
      canonicalToAdopted,
      _routerInfo
    );
```

Would save us an argument in general and some gas in the case that this is a slow liq transfer